File: app/src/test/java/com/YucelDigital/ilactakip/MedicineTest.java
package com.YucelDigital.ilactakip;

import org.junit.Test;
import org.junit.Before;
import static org.junit.Assert.*;

import java.util.HashMap;
import java.util.Calendar;

/**
 * Medicine model sınıfı için unit testler.
 * Constructor varsayılan değerleri, getter/setter doğruluğu,
 * güne özel zamanlama ve Serializable uyumluluğunu test eder.
 */
public class MedicineTest {

    private Medicine medicine;

    @Before
    public void setUp() {
        medicine = new Medicine("Parol", "08:00", "1 Oca - 31 Oca", "Tok karnına");
    }

    // ══════════════════════════════════════════════════════════════
    //  Constructor Varsayılan Değerleri
    // ══════════════════════════════════════════════════════════════

    @Test
    public void constructor_setsNameCorrectly() {
        assertEquals("Parol", medicine.getName());
    }

    @Test
    public void constructor_setsTimeCorrectly() {
        assertEquals("08:00", medicine.getTime());
    }

    @Test
    public void constructor_setsDateRangeCorrectly() {
        assertEquals("1 Oca - 31 Oca", medicine.getDateRange());
    }

    @Test
    public void constructor_setsNoteCorrectly() {
        assertEquals("Tok karnına", medicine.getNote());
    }

    @Test
    public void constructor_defaultsActiveToTrue() {
        assertTrue(medicine.isActive());
    }

    @Test
    public void constructor_defaultsTakenToFalse() {
        assertFalse(medicine.isTaken());
    }

    @Test
    public void constructor_defaultsStartDateToZero() {
        assertEquals(0, medicine.getStartDate());
    }

    @Test
    public void constructor_defaultsEndDateToZero() {
        assertEquals(0, medicine.getEndDate());
    }

    @Test
    public void constructor_defaultsIntervalDaysToOne() {
        assertEquals(1, medicine.getIntervalDays());
    }

    @Test
    public void constructor_defaultsSoundUriToNull() {
        assertNull(medicine.getSoundUri());
    }

    @Test
    public void constructor_defaultsUseCustomDaysToFalse() {
        assertFalse(medicine.isUseCustomDays());
    }

    @Test
    public void constructor_defaultsCustomDayTimesToNull() {
        assertNull(medicine.getCustomDayTimes());
    }

    // ══════════════════════════════════════════════════════════════
    //  Setter / Getter Doğruluğu
    // ══════════════════════════════════════════════════════════════

    @Test
    public void setActive_false_isActiveReturnsFalse() {
        medicine.setActive(false);
        assertFalse(medicine.isActive());
    }

    @Test
    public void setActive_true_isActiveReturnsTrue() {
        medicine.setActive(false);
        medicine.setActive(true);
        assertTrue(medicine.isActive());
    }

    @Test
    public void setTaken_true_isTakenReturnsTrue() {
        medicine.setTaken(true);
        assertTrue(medicine.isTaken());
    }

    @Test
    public void setTaken_false_isTakenReturnsFalse() {
        medicine.setTaken(true);
        medicine.setTaken(false);
        assertFalse(medicine.isTaken());
    }

    @Test
    public void setTime_updatesTime() {
        medicine.setTime("14:00");
        assertEquals("14:00", medicine.getTime());
    }

    @Test
    public void setStartDate_updatesStartDate() {
        long date = System.currentTimeMillis();
        medicine.setStartDate(date);
        assertEquals(date, medicine.getStartDate());
    }

    @Test
    public void setEndDate_updatesEndDate() {
        long date = System.currentTimeMillis();
        medicine.setEndDate(date);
        assertEquals(date, medicine.getEndDate());
    }

    @Test
    public void setIntervalDays_updatesIntervalDays() {
        medicine.setIntervalDays(3);
        assertEquals(3, medicine.getIntervalDays());
    }

    @Test
    public void setSoundUri_updatesSoundUri() {
        medicine.setSoundUri("content://media/alarm/1");
        assertEquals("content://media/alarm/1", medicine.getSoundUri());
    }

    @Test
    public void setSoundUri_null_returnsNull() {
        medicine.setSoundUri("content://media/alarm/1");
        medicine.setSoundUri(null);
        assertNull(medicine.getSoundUri());
    }

    // ══════════════════════════════════════════════════════════════
    //  Güne Özel Zamanlama
    // ══════════════════════════════════════════════════════════════

    @Test
    public void setUseCustomDays_true_isUseCustomDaysReturnsTrue() {
        medicine.setUseCustomDays(true);
        assertTrue(medicine.isUseCustomDays());
    }

    @Test
    public void setCustomDayTimes_setsCorrectlyAndReturns() {
        HashMap<Integer, String> dayTimes = new HashMap<>();
        dayTimes.put(Calendar.MONDAY, "09:00");
        dayTimes.put(Calendar.FRIDAY, "14:00");

        medicine.setCustomDayTimes(dayTimes);

        assertNotNull(medicine.getCustomDayTimes());
        assertEquals(2, medicine.getCustomDayTimes().size());
        assertEquals("09:00", medicine.getCustomDayTimes().get(Calendar.MONDAY));
        assertEquals("14:00", medicine.getCustomDayTimes().get(Calendar.FRIDAY));
    }

    @Test
    public void customDayTimes_doesNotContainUnsetDay() {
        HashMap<Integer, String> dayTimes = new HashMap<>();
        dayTimes.put(Calendar.MONDAY, "09:00");
        medicine.setCustomDayTimes(dayTimes);

        assertFalse(medicine.getCustomDayTimes().containsKey(Calendar.WEDNESDAY));
    }

    @Test
    public void setCustomDayTimes_null_returnsNull() {
        medicine.setCustomDayTimes(new HashMap<>());
        medicine.setCustomDayTimes(null);
        assertNull(medicine.getCustomDayTimes());
    }

    @Test
    public void customDayTimes_allDaysOfWeek() {
        HashMap<Integer, String> dayTimes = new HashMap<>();
        dayTimes.put(Calendar.MONDAY,    "07:00");
        dayTimes.put(Calendar.TUESDAY,   "08:00");
        dayTimes.put(Calendar.WEDNESDAY, "09:00");
        dayTimes.put(Calendar.THURSDAY,  "10:00");
        dayTimes.put(Calendar.FRIDAY,    "11:00");
        dayTimes.put(Calendar.SATURDAY,  "12:00");
        dayTimes.put(Calendar.SUNDAY,    "13:00");

        medicine.setCustomDayTimes(dayTimes);
        assertEquals(7, medicine.getCustomDayTimes().size());
    }

    @Test
    public void customDayTimes_multipleTimesPerDay_commaFormat() {
        HashMap<Integer, String> dayTimes = new HashMap<>();
        dayTimes.put(Calendar.MONDAY, "08:00, 14:00, 20:00");
        medicine.setCustomDayTimes(dayTimes);

        String times = medicine.getCustomDayTimes().get(Calendar.MONDAY);
        assertNotNull(times);
        assertTrue(times.contains(","));
        assertEquals(3, times.split(",").length);
    }

    // ══════════════════════════════════════════════════════════════
    //  Çoklu Saat Formatı (Standart Mod)
    // ══════════════════════════════════════════════════════════════

    @Test
    public void time_withMultipleTimesComma_splitCorrectly() {
        Medicine multi = new Medicine("Aspirin", "08:00, 14:00, 20:00", "", "");
        String[] times = multi.getTime().split(", ");
        assertEquals(3, times.length);
        assertEquals("08:00", times[0]);
        assertEquals("14:00", times[1]);
        assertEquals("20:00", times[2]);
    }

    @Test
    public void time_singleTime_noCommaSplit() {
        String[] times = medicine.getTime().split(", ");
        assertEquals(1, times.length);
        assertEquals("08:00", times[0]);
    }

    // ══════════════════════════════════════════════════════════════
    //  Boş / Null Değerler
    // ══════════════════════════════════════════════════════════════

    @Test
    public void constructor_emptyStrings_noException() {
        Medicine m = new Medicine("", "", "", "");
        assertEquals("", m.getName());
        assertEquals("", m.getTime());
        assertEquals("", m.getDateRange());
        assertEquals("", m.getNote());
    }

    @Test
    public void constructor_nullNote_isNull() {
        Medicine m = new Medicine("Test", "10:00", null, null);
        assertNull(m.getNote());
        assertNull(m.getDateRange());
    }

    // ══════════════════════════════════════════════════════════════
    //  Serializable Uyumluluğu
    // ══════════════════════════════════════════════════════════════

    @Test
    public void medicine_implementsSerializable() {
        assertTrue(medicine instanceof java.io.Serializable);
    }

    @Test
    public void medicine_serialization_roundTrip() throws Exception {
        // Tüm alanları doldur
        medicine.setActive(false);
        medicine.setTaken(true);
        medicine.setStartDate(1000L);
        medicine.setEndDate(2000L);
        medicine.setIntervalDays(3);
        medicine.setSoundUri("content://alarm/1");
        medicine.setUseCustomDays(true);
        HashMap<Integer, String> dayTimes = new HashMap<>();
        dayTimes.put(Calendar.MONDAY, "09:00");
        medicine.setCustomDayTimes(dayTimes);

        // Serialize
        java.io.ByteArrayOutputStream bos = new java.io.ByteArrayOutputStream();
        java.io.ObjectOutputStream oos = new java.io.ObjectOutputStream(bos);
        oos.writeObject(medicine);
        oos.close();

        // Deserialize
        java.io.ByteArrayInputStream bis = new java.io.ByteArrayInputStream(bos.toByteArray());
        java.io.ObjectInputStream ois = new java.io.ObjectInputStream(bis);
        Medicine deserialized = (Medicine) ois.readObject();
        ois.close();

        // Doğrulama
        assertEquals(medicine.getName(), deserialized.getName());
        assertEquals(medicine.getTime(), deserialized.getTime());
        assertEquals(medicine.getDateRange(), deserialized.getDateRange());
        assertEquals(medicine.getNote(), deserialized.getNote());
        assertEquals(medicine.isActive(), deserialized.isActive());
        assertEquals(medicine.isTaken(), deserialized.isTaken());
        assertEquals(medicine.getStartDate(), deserialized.getStartDate());
        assertEquals(medicine.getEndDate(), deserialized.getEndDate());
        assertEquals(medicine.getIntervalDays(), deserialized.getIntervalDays());
        assertEquals(medicine.getSoundUri(), deserialized.getSoundUri());
        assertEquals(medicine.isUseCustomDays(), deserialized.isUseCustomDays());
        assertNotNull(deserialized.getCustomDayTimes());
        assertEquals("09:00", deserialized.getCustomDayTimes().get(Calendar.MONDAY));
    }

    // ══════════════════════════════════════════════════════════════
    //  Alarm ID Hesaplama Tutarlılığı
    // ══════════════════════════════════════════════════════════════

    @Test
    public void alarmId_standardMode_isConsistent() {
        // AlarmId = (name + time).hashCode()
        String name = "Parol";
        String time = "08:00";
        int id1 = (name + time).hashCode();
        int id2 = (name + time).hashCode();
        assertEquals(id1, id2);
    }

    @Test
    public void alarmId_customDay_isConsistent() {
        // AlarmId = (name + "_day" + calDay + "_" + time).hashCode()
        String name = "Parol";
        int calDay = Calendar.MONDAY;
        String time = "09:00";
        int id1 = (name + "_day" + calDay + "_" + time).hashCode();
        int id2 = (name + "_day" + calDay + "_" + time).hashCode();
        assertEquals(id1, id2);
    }

    @Test
    public void alarmId_differentDays_areDifferent() {
        String name = "Parol";
        String time = "09:00";
        int idMon = (name + "_day" + Calendar.MONDAY + "_" + time).hashCode();
        int idFri = (name + "_day" + Calendar.FRIDAY + "_" + time).hashCode();
        assertNotEquals(idMon, idFri);
    }

    @Test
    public void alarmId_differentNames_areDifferent() {
        String time = "08:00";
        int id1 = ("Parol" + time).hashCode();
        int id2 = ("Aspirin" + time).hashCode();
        assertNotEquals(id1, id2);
    }

    @Test
    public void alarmId_differentTimes_areDifferent() {
        String name = "Parol";
        int id1 = (name + "08:00").hashCode();
        int id2 = (name + "14:00").hashCode();
        assertNotEquals(id1, id2);
    }

    @Test
    public void resetAlarmId_isDifferentFromMainAlarmId() {
        String name = "Parol";
        String time = "08:00";
        int mainId  = (name + time).hashCode();
        int resetId = (name + time + "_reset").hashCode();
        assertNotEquals(mainId, resetId);
    }
}

File: app/src/test/java/com/YucelDigital/ilactakip/MedicineTest.kt
package com.YucelDigital.ilactakip

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNotSame
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

import java.util.Calendar

/**
 * Medicine model sınıfı için unit testler.
 * Constructor varsayılan değerleri, getter/setter doğruluğu,
 * güne özel zamanlama ve Serializable uyumluluğunu test eder.
 */
class MedicineTest {

    private lateinit var medicine: Medicine

    @Before
    fun setUp() {
        medicine = Medicine("Parol", "08:00", "1 Oca - 31 Oca", "Tok karnına")
    }

    // ══════════════════════════════════════════════════════════════
    //  Constructor Varsayılan Değerleri
    // ══════════════════════════════════════════════════════════════

    @Test
    fun constructor_setsNameCorrectly() {
        assertEquals("Parol", medicine.name)
    }

    @Test
    fun constructor_setsTimeCorrectly() {
        assertEquals("08:00", medicine.time)
    }

    @Test
    fun constructor_setsDateRangeCorrectly() {
        assertEquals("1 Oca - 31 Oca", medicine.dateRange)
    }

    @Test
    fun constructor_setsNoteCorrectly() {
        assertEquals("Tok karnına", medicine.note)
    }

    @Test
    fun constructor_defaultsActiveToTrue() {
        assertTrue(medicine.isActive)
    }

    @Test
    fun constructor_defaultsTakenToFalse() {
        assertFalse(medicine.isTaken)
    }

    @Test
    fun constructor_defaultsStartDateToZero() {
        assertEquals(0L, medicine.startDate)
    }

    @Test
    fun constructor_defaultsEndDateToZero() {
        assertEquals(0L, medicine.endDate)
    }

    @Test
    fun constructor_defaultsIntervalDaysToOne() {
        assertEquals(1, medicine.intervalDays)
    }

    @Test
    fun constructor_defaultsSoundUriToNull() {
        assertNull(medicine.soundUri)
    }

    @Test
    fun constructor_defaultsUseCustomDaysToFalse() {
        assertFalse(medicine.isUseCustomDays)
    }

    @Test
    fun constructor_defaultsCustomDayTimesToNull() {
        assertNull(medicine.customDayTimes)
    }

    // ══════════════════════════════════════════════════════════════
    //  Setter / Getter Doğruluğu
    // ══════════════════════════════════════════════════════════════

    @Test
    fun setActive_false_isActiveReturnsFalse() {
        medicine.isActive = false
        assertFalse(medicine.isActive)
    }

    @Test
    fun setActive_true_isActiveReturnsTrue() {
        medicine.isActive = false
        medicine.isActive = true
        assertTrue(medicine.isActive)
    }

    @Test
    fun setTaken_true_isTakenReturnsTrue() {
        medicine.isTaken = true
        assertTrue(medicine.isTaken)
    }

    @Test
    fun setTaken_false_isTakenReturnsFalse() {
        medicine.isTaken = true
        medicine.isTaken = false
        assertFalse(medicine.isTaken)
    }

    @Test
    fun setTime_updatesTime() {
        medicine.time = "14:00"
        assertEquals("14:00", medicine.time)
    }

    @Test
    fun setStartDate_updatesStartDate() {
        val date = System.currentTimeMillis()
        medicine.startDate = date
        assertEquals(date, medicine.startDate)
    }

    @Test
    fun setEndDate_updatesEndDate() {
        val date = System.currentTimeMillis()
        medicine.endDate = date
        assertEquals(date, medicine.endDate)
    }

    @Test
    fun setIntervalDays_updatesIntervalDays() {
        medicine.intervalDays = 3
        assertEquals(3, medicine.intervalDays)
    }

    @Test
    fun setSoundUri_updatesSoundUri() {
        medicine.soundUri = "content://media/alarm/1"
        assertEquals("content://media/alarm/1", medicine.soundUri)
    }

    @Test
    fun setSoundUri_null_returnsNull() {
        medicine.soundUri = "content://media/alarm/1"
        medicine.soundUri = null
        assertNull(medicine.soundUri)
    }

    // ══════════════════════════════════════════════════════════════
    //  Güne Özel Zamanlama
    // ══════════════════════════════════════════════════════════════

    @Test
    fun setUseCustomDays_true_isUseCustomDaysReturnsTrue() {
        medicine.isUseCustomDays = true
        assertTrue(medicine.isUseCustomDays)
    }

    @Test
    fun setCustomDayTimes_setsCorrectlyAndReturns() {
        val dayTimes = HashMap<Int, String>()
        dayTimes[Calendar.MONDAY] = "09:00"
        dayTimes[Calendar.FRIDAY] = "14:00"

        medicine.customDayTimes = dayTimes

        assertNotNull(medicine.customDayTimes)
        assertEquals(2, medicine.customDayTimes!!.size)
        assertEquals("09:00", medicine.customDayTimes!![Calendar.MONDAY])
        assertEquals("14:00", medicine.customDayTimes!![Calendar.FRIDAY])
    }

    @Test
    fun customDayTimes_doesNotContainUnsetDay() {
        val dayTimes = HashMap<Int, String>()
        dayTimes[Calendar.MONDAY] = "09:00"
        medicine.customDayTimes = dayTimes

        assertFalse(medicine.customDayTimes!!.containsKey(Calendar.WEDNESDAY))
    }

    @Test
    fun setCustomDayTimes_null_returnsNull() {
        medicine.customDayTimes = HashMap()
        medicine.customDayTimes = null
        assertNull(medicine.customDayTimes)
    }

    @Test
    fun customDayTimes_allDaysOfWeek() {
        val dayTimes = HashMap<Int, String>()
        dayTimes[Calendar.MONDAY] = "07:00"
        dayTimes[Calendar.TUESDAY] = "08:00"
        dayTimes[Calendar.WEDNESDAY] = "09:00"
        dayTimes[Calendar.THURSDAY] = "10:00"
        dayTimes[Calendar.FRIDAY] = "11:00"
        dayTimes[Calendar.SATURDAY] = "12:00"
        dayTimes[Calendar.SUNDAY] = "13:00"

        medicine.customDayTimes = dayTimes
        assertEquals(7, medicine.customDayTimes!!.size)
    }

    @Test
    fun customDayTimes_multipleTimesPerDay_commaFormat() {
        val dayTimes = HashMap<Int, String>()
        dayTimes[Calendar.MONDAY] = "08:00, 14:00, 20:00"
        medicine.customDayTimes = dayTimes

        val times = medicine.customDayTimes!![Calendar.MONDAY]
        assertNotNull(times)
        assertTrue(times!!.contains(","))
        assertEquals(3, times.split(",").size)
    }

    // ══════════════════════════════════════════════════════════════
    //  Çoklu Saat Formatı (Standart Mod)
    // ══════════════════════════════════════════════════════════════

    @Test
    fun time_withMultipleTimesComma_splitCorrectly() {
        val multi = Medicine("Aspirin", "08:00, 14:00, 20:00", "", "")
        val times = multi.time!!.split(", ")
        assertEquals(3, times.size)
        assertEquals("08:00", times[0])
        assertEquals("14:00", times[1])
        assertEquals("20:00", times[2])
    }

    @Test
    fun time_singleTime_noCommaSplit() {
        val times = medicine.time!!.split(", ")
        assertEquals(1, times.size)
        assertEquals("08:00", times[0])
    }

    // ══════════════════════════════════════════════════════════════
    //  Boş / Null Değerler
    // ══════════════════════════════════════════════════════════════

    @Test
    fun constructor_emptyStrings_noException() {
        val m = Medicine("", "", "", "")
        assertEquals("", m.name)
        assertEquals("", m.time)
        assertEquals("", m.dateRange)
        assertEquals("", m.note)
    }

    @Test
    fun constructor_nullNote_isNull() {
        val m = Medicine("Test", "10:00", null, null)
        assertNull(m.note)
        assertNull(m.dateRange)
    }

    // ══════════════════════════════════════════════════════════════
    //  Serializable Uyumluluğu
    // ══════════════════════════════════════════════════════════════

    @Test
    fun medicine_implementsSerializable() {
        assertTrue(medicine is java.io.Serializable)
    }

    @Test
    fun medicine_serialization_roundTrip() {
        // Tüm alanları doldur
        medicine.isActive = false
        medicine.isTaken = true
        medicine.startDate = 1000L
        medicine.endDate = 2000L
        medicine.intervalDays = 3
        medicine.soundUri = "content://alarm/1"
        medicine.isUseCustomDays = true
        val dayTimes = HashMap<Int, String>()
        dayTimes[Calendar.MONDAY] = "09:00"
        medicine.customDayTimes = dayTimes

        // Serialize
        val bos = java.io.ByteArrayOutputStream()
        val oos = java.io.ObjectOutputStream(bos)
        oos.writeObject(medicine)
        oos.close()

        // Deserialize
        val bis = java.io.ByteArrayInputStream(bos.toByteArray())
        val ois = java.io.ObjectInputStream(bis)
        val deserialized = ois.readObject() as Medicine
        ois.close()

        // Doğrulama
        assertEquals(medicine.name, deserialized.name)
        assertEquals(medicine.time, deserialized.time)
        assertEquals(medicine.dateRange, deserialized.dateRange)
        assertEquals(medicine.note, deserialized.note)
        assertEquals(medicine.isActive, deserialized.isActive)
        assertEquals(medicine.isTaken, deserialized.isTaken)
        assertEquals(medicine.startDate, deserialized.startDate)
        assertEquals(medicine.endDate, deserialized.endDate)
        assertEquals(medicine.intervalDays, deserialized.intervalDays)
        assertEquals(medicine.soundUri, deserialized.soundUri)
        assertEquals(medicine.isUseCustomDays, deserialized.isUseCustomDays)
        assertNotNull(deserialized.customDayTimes)
        assertEquals("09:00", deserialized.customDayTimes!![Calendar.MONDAY])
    }

    // ══════════════════════════════════════════════════════════════
    //  Alarm ID Hesaplama Tutarlılığı
    // ══════════════════════════════════════════════════════════════

    @Test
    fun alarmId_standardMode_isConsistent() {
        // AlarmId = (name + time).hashCode()
        val name = "Parol"
        val time = "08:00"
        val id1 = (name + time).hashCode()
        val id2 = (name + time).hashCode()
        assertEquals(id1, id2)
    }

    @Test
    fun alarmId_customDay_isConsistent() {
        // AlarmId = (name + "_day" + calDay + "_" + time).hashCode()
        val name = "Parol"
        val calDay = Calendar.MONDAY
        val time = "09:00"
        val id1 = (name + "_day" + calDay + "_" + time).hashCode()
        val id2 = (name + "_day" + calDay + "_" + time).hashCode()
        assertEquals(id1, id2)
    }

    @Test
    fun alarmId_differentDays_areDifferent() {
        val name = "Parol"
        val time = "09:00"
        val idMon = (name + "_day" + Calendar.MONDAY + "_" + time).hashCode()
        val idFri = (name + "_day" + Calendar.FRIDAY + "_" + time).hashCode()
        assertNotEquals(idMon, idFri)
    }

    @Test
    fun alarmId_differentNames_areDifferent() {
        val time = "08:00"
        val id1 = ("Parol" + time).hashCode()
        val id2 = ("Aspirin" + time).hashCode()
        assertNotEquals(id1, id2)
    }

    @Test
    fun alarmId_differentTimes_areDifferent() {
        val name = "Parol"
        val id1 = (name + "08:00").hashCode()
        val id2 = (name + "14:00").hashCode()
        assertNotEquals(id1, id2)
    }

    @Test
    fun resetAlarmId_isDifferentFromMainAlarmId() {
        val name = "Parol"
        val time = "08:00"
        val mainId = (name + time).hashCode()
        val resetId = (name + time + "_reset").hashCode()
        assertNotEquals(mainId, resetId)
    }

    // ══════════════════════════════════════════════════════════════
    //  copy() — Compose'da yeniden çizim için kullanılan klon
    // ══════════════════════════════════════════════════════════════

    @Test
    fun copy_producesDistinctInstance() {
        val copy = medicine.copy()
        assertNotSame(medicine, copy)
    }

    @Test
    fun copy_copiesAllScalarFields() {
        medicine.isActive = false
        medicine.isTaken = true
        medicine.startDate = 1000L
        medicine.endDate = 2000L
        medicine.intervalDays = 3
        medicine.soundUri = "content://alarm/1"
        medicine.isUseCustomDays = true

        val copy = medicine.copy()

        assertEquals(medicine.name, copy.name)
        assertEquals(medicine.time, copy.time)
        assertEquals(medicine.dateRange, copy.dateRange)
        assertEquals(medicine.note, copy.note)
        assertEquals(medicine.isActive, copy.isActive)
        assertEquals(medicine.isTaken, copy.isTaken)
        assertEquals(medicine.startDate, copy.startDate)
        assertEquals(medicine.endDate, copy.endDate)
        assertEquals(medicine.intervalDays, copy.intervalDays)
        assertEquals(medicine.soundUri, copy.soundUri)
        assertEquals(medicine.isUseCustomDays, copy.isUseCustomDays)
    }

    @Test
    fun copy_customDayTimes_isDeepCopied() {
        val dayTimes = HashMap<Int, String>()
        dayTimes[Calendar.MONDAY] = "09:00"
        medicine.customDayTimes = dayTimes

        val copy = medicine.copy()

        // İçerik aynı ama harita farklı bir örnek (birini değiştirmek diğerini etkilemesin)
        assertNotSame(medicine.customDayTimes, copy.customDayTimes)
        assertEquals("09:00", copy.customDayTimes!![Calendar.MONDAY])

        copy.customDayTimes!![Calendar.TUESDAY] = "10:00"
        assertFalse(medicine.customDayTimes!!.containsKey(Calendar.TUESDAY))
    }

    @Test
    fun copy_nullCustomDayTimes_staysNull() {
        assertNull(medicine.customDayTimes)
        assertNull(medicine.copy().customDayTimes)
    }

    @Test
    fun copy_thenMutate_doesNotAffectOriginal() {
        val original = Medicine("Parol", "08:00", "", "")
        original.isTaken = false

        val updated = original.copy().apply { isTaken = true }

        assertTrue(updated.isTaken)
        assertFalse("Klonu değiştirmek orijinali etkilememeli", original.isTaken)
    }
}
